### DISCUSSION TITLE
Extension State after Pylance Changes

### DISCUSSION THREAD

**Discussion by nickjmiller** [nickjmiller/pytest-fixtures#47] [Announcements]
Going to collect thoughts here on keeping the extension on the marketplace now that pylance has extended support to pytest fixtures.

It seems like pylance's fixture support will be much faster, has more complete type annotations, and native support for additional features (such as Refactor->Rename).

pytest-fixtures uses  `pytest --fixtures` directly, so it may be more true to the test  environment, and maybe more customizable with `additionalArgs`.

Pylance is closed source, some users opt not to use it because of that. Specifically, [open vsx registry](https://open-vsx.org/extension/nickmillerdev/pytest-fixtures).

@jxramos you had thoughts as well, could you list them here?

**Comment by jxramos** [comment]
Yah, we tried Pylance at work after disabling pytest-fixtures and witnessed it jumping to irrelevant fixtures and also in some scenarios we’ve yet to characterize Pylance was unable to do anything and didn’t respond to cmd clicks in macOS. There are however some extenuating circumstances I’m working through that involve generated code from a bazel workspace that may be complicating things.

I would think that even if Pylance indexed the whole repo it would still lack the discriminating abilities that pytest can leverage in its ability to specify pytest plug-ins with the whole -p argument inputs. That’s input that’s not reflected in the python spec and I imagine has no equivalent.


Interestingly enough we found pycharm able to jump to some more fixture definitions but had inaccuracies of its own. This was all evaluated in a cursory spot checking evaluation, nothing too scientific or well documented along the way just a brief pair programming session when me and a coworker messed around scouting capabilities.

**Comment by jxramos** [comment]
I went snooping around to study the `--fixtures` option that pytest makes available. I'm somewhat interested in seeing if the pytest developers would be interested in exporting a json file containing the same information which can be leveraged to tie into the vscode language server stuff to facilitate navigational operations and populate hovertool stuff with.

We know that this pytest-fixtures extension parses the output from `--fixtures` over in this piece of logic https://github.com/nickjmiller/pytest-fixtures/blob/fd51337429da0faa67ddd5414d684e56323854b1/src/fixtureParser.ts#L62

so if there existed a serialized form of that same fixture information we could dodge the parsing and get right to loading the location data etc straight from the json file. The only thing I'm uncertain of is whether or not vscode has permissions to write temporary files to disk and what not. Would you happen to know those details @nickjmiller?

# Fixtures Pytest Mode Implementation
Looking at the pytest repo here is where the command line argument definition is found for `--fixtures`, noting its destination `dest=showfixtures`
https://github.com/pytest-dev/pytest/blob/4546d5445aaefe6a03957db028c263521dfb5c4b/src/_pytest/python.py#L89-L97

And this is where `showfixtures` gets backed by its actual implementation and does the colorized terminal writer operations with all the found fixture data.
https://github.com/pytest-dev/pytest/blob/4546d5445aaefe6a03957db028c263521dfb5c4b/src/_pytest/python.py#L133-L136
https://github.com/pytest-dev/pytest/blob/4546d5445aaefe6a03957db028c263521dfb5c4b/src/_pytest/python.py#L1593-L1651

So pytest first executes a collection phase operation with  `session.perform_collect()`, which guarantees that the session's fixture manager (`session._fixturemanager`) will be fully populated with all collected fixtures on hand according to all the regular pytest collection discovery mechanisms and rules used by the app itself under normal operation. Afterwards the fixture manager's internal  dict `_arg2fixturedefs` is iterated over to collate the fixture data into a list of tuples which gets sorted. And that sorted list gets printed where `showfixtures` writes the following fixture data.
1. The name of the fixture
2. The scope of the fixture if changed from the default
3. The path to the fixture's definition in its source module
4. The docstring of the fixture if found

# Possible Json Output
It should be pretty straightforward to pull this off, the json would basically be based off some derivative dict from `_arg2fixturedefs` that can be directly rendered to json with some appropriately named keys for the fields. It would be a straight list of dicts looking something like

```json
[
  {
    "name": "fixture1",
    "scope": "session",
    "path": "/projX/module_plugin_z.py"
    "docstring": "..."
  },
  {
    "name": "fixture2",
    "scope": "module",
    "path": "/projX/module_plugin_k.py"
    "docstring": "..."
  },
  {
    "name": "fixture3",
    "scope": "function",
    "path": "/projX/module_plugin_k.py"
    "docstring": "..."
  }
]
```

**Reply by jxramos** [reply]
I attained getting the json output working for pytest in this forked version of the project https://github.com/jxramos/pytest/commit/4473dd96242de596131f3d9a049c406d32d8b6a1. 


Here's some sample output from a dummy test project I made in the forked pytest repo: [fixture_data.json](https://github.com/nickjmiller/pytest-fixtures/files/14132500/fixture_data.json)

```
$ tree test_proj
test_proj
├── conftest.py
├── plugins
│   └── plugin_z.py
└── test_module.py
```

with 

```py
#========================================
# test_proj/conftest.py
#========================================
import pytest

@pytest.fixture(scope="session")
def session_scope_fix():
    return 4


@pytest.fixture(scope="function")
def func_scope_fix():
    return 0

#========================================
# test_proj/test_module.py
#========================================
import pytest

@pytest.fixture
def file_scope_fix():
    return "a"

#========================================
#test_proj/plugins/plugin_z.py
#========================================
import pytest

@pytest.fixture(scope="module")
def mod_scope_fix():
    return 3
```

I'm imagining the vscode extension could pivot to this functionality if accepted by the pytest project. Otherwise it could be packaged as a third party pytest plugin if they don't feel it's suitable for the core project. If the latter works out being the case then this extension could take a configuration to switch modes for fixture discovery from parsing to the json file style.

Then once armed with the json output the extension could retain that information in memory and refresh it with each test module that is navigated upon. I believe a test module reference state would need to be maintained to support second and third order hops away from the test module as one drills down across fixtures that jump around the project into conftest and plugin files whereby the `--fixtures` command would not have the proper scope to orient itself on the python modules themselves. That's the limitation of `--fixtures` being invoked per file navigation event, the variety of files you can land upon simply are not all equal with respect to fixture visibility.

To be specific, I'm referring to this logic here that exercises the fixtures discovery on the active file `document.uri.fsPath`
https://github.com/nickjmiller/pytest-fixtures/blob/fd51337429da0faa67ddd5414d684e56323854b1/src/fixture.ts#L41-L43

As you jump file to file not every file has the full context of the project and you'll wind up in navigational dead ends as different fixtures are visible to different degrees depending on what location you're at. For example if you're off in some distant plugin folder there's no immediate context to bring the full session scope into existence. There's a few things in tension with the current design but I think the gem of the idea you had to use the pytest app itself to communicate its known fixtures is a real winner. Everything else in this vscode extension will be keeping the context straight as you jump file to file drilling down across the fixture dependencies which in my project at work can span 5 fixtures deep.

**Comment by jxramos** [comment]
Sounds like it is possible to read and write json from a vscode extension https://stackoverflow.com/a/51328232/1330381

**Comment by nickjmiller** [comment]
Thanks @jxramos , I think this is a good idea and I'd love to get rid of the fragile string matching we are doing right now. If this is merged I'd be happy to update it or accept a pull request, I think regardless this should be a toggleable option since we don't want to break the extension for people using old versions of pytest.

**Reply by jxramos** [reply]
I just had an epiphany regarding the satisfaction of long distance fixture definitions relative to the active file (ie `document.uri.fsPath`). I've long realized the inclusion of all plugins was a fairly straightforward proposition via the `pytest-fixtures.extraArguments` route, you simply enumerate the plugins in scope for a given test project. However I was getting tripped up on wrestling with how to retain the fixtures in conftest files so they're present to any active file.

What I experimented with this afternoon was passing a stack of conftest references up the directory tree however deep the test project demanded, eg `-p test_proj.xyz.conftest -p test_proj.conftest -p conftest`. I wasn't sure how pytest would take being given a plugin that was actually a conftest file. Pytest runs and executes happily which is great news, and the fixtures I created in my dummy project do appear in the results for `--fixtures` output. I'm not yet sure of the full ramifications of scoping between different subtrees when fixtures come into name collisions and whether or not my *conftest as plugin input technique* would thwart some of the true fidelity used in pytest in fixture resolution during the collection phase. I may be artificially tampering with the precedence rules using my hack. But for those fixtures where name collisions are not expected between conftest files found in sibling sub test directories this technique could work really great.

**Reply by jxramos** [reply]
had another epiphany today, regarding the worry about the ramifications of mixing plugins and conftests as plugins, so the fixture definition precedence resolution logic should be near equal if not identical to pytest's so long as the conftests are kept at the end of the arguments list, so basically something like this...
```
-p projectA.dirA.plugin_a
-p projectB.plugin_b
...
-p conftest
-p test_foo.conftest
-p test_foo.test_bar.conftest
```

I manage producing such a list in bazel with two separate bazel queries, the first pulling all the plugins with tags query after having first tagged all plugin targets with an identical keyword and the second query by grabbing those targets with have `scrs == ["conftest.py"]`.
